ext/include/opentelemetry/ext/http/client/http_client_factory.h

```diff
@@ -0,0 +1,27 @@
+#pragma once
+#include "opentelemetry/ext/http/client/http_client.h"
+#ifdef WITH_CURL
+#  include "opentelemetry/ext/http/client/curl/http_client_curl.h"
+#endif
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace ext
+{
+namespace http
+{
+namespace client
+{
+class HttpClientFactory
+{
+public:
+#ifdef WITH_CURL
+  static std::shared_ptr<SessionManager> Create()
```

Comment:
Sorry for belated response: my thinking here is that the `http_client_factory.h` itself may remain abstract. Any client that implements this interface must implement all generic methods. Or provide a no-op method in case if some method is not applicable to specific HTTP client.. Basically, what I think looks wrong is that we have `#ifdef` in the interface specification. It should not need any ifdefs.

What is the full set of methods you'd need for the `HttpClientFactory` interface? If `SessionManager` is unique to curl, should this method be moved to `CurlHttpClient` rather than adding ifdef here?

I'd like to refer to these interfaces as example:
https://github.com/microsoft/cpp_client_telemetry/blob/master/lib/http/HttpClientFactory.hpp
https://github.com/microsoft/cpp_client_telemetry/blob/master/lib/include/public/IHttpClient.hpp

This code is generic with 6 different implementations cross-plat, no ifdefs.

Reply:
Nice to see reference of `cpp_client_telemetry` now that it is public  :)
So, session Manager is not unique to `curl` but an abstract class which is implemented for all client libraries. 
`opentelmetry::ext::http::client::SessionManager` is similar to `MAT_NS_BEGIN::IHttpClient`, and it manages multiple ongoing HTTP requests.
`opentelementry::ext::http::client::Session` logically represent a (possibly) ongoing http request.

http client factory implementation is similar to what is done in `cpp_client_telemetry` as here:
https://github.com/microsoft/cpp_client_telemetry/blob/master/lib/http/HttpClientFactory.cpp

In above, we have #ifdef for each client implementation.

Reply:
Please see the example of the http request flow here: https://github.com/open-telemetry/opentelemetry-cpp/blob/aa7b3c8f17918c71785ff9213d7e1f58c0de3867/ext/include/opentelemetry/ext/http/client/http_client.h#L11

```
  SessionManager sessionManager; // implementer can provide singleton implementation for it
  auto session = sessionManager.createSession("localhost", 8000);
  auto request = session->CreateRequest();
  request->AddHeader(..);
  SimpleResponseHandler res_handler;
  session->SendRequest(res_handler);
  session->FinishSession() // optionally in the end
  ...shutdown
  sessionManager.FinishAllSessions()
```